### PR TITLE
Refactoring/Genericising to add contributions card payment flow

### PR DESCRIPTION
**Pull request by twrichards**
Some renaming/refactoring/restructuring/genericising to allow for contributions card payment update flow (also reduced some package tangle). Also added a basic contributions tab (full parity with profile will follow in a future PR).

### CODE REVIEW

File: app/client/components/payment/update/paymentUpdated.tsx [hunk 1]
```diff
@@ -64,15 +68,23 @@ const ConfirmedNewPaymentDetailsRenderer = (subscription: Subscription) => {
...
     <h2>
-      Thank you. You are helping to support independent investigative journalism
+      Thank you. You are helping to support independent investigative
+      journalism.
     </h2>
     <div>
-      <LinkButton to="/" text="Manage your account" primary right />
+      <LinkButton
+        to={"/" + productType.urlPart}
+        text={"Manage your " + productType.friendlyName}
```
Comment (paulbrown1982) [inline app/client/components/payment/update/paymentUpdated.tsx]: Just an interesting one for the record: when we onboard Voucher and Home Delivery, they will be seen as one product/tab e.g. "Newspaper Subscription", with a page that must variate appropriately.

File: app/client/components/payment/update/stripeCardInputForm.tsx [hunk 2]
```diff
@@ -68,54 +68,59 @@ export class StripeCardInputForm extends React.Component<
...
+                display: "flex",
+                flexWrap: "wrap",
+                justifyContent: "space-between"
+              }}
+            >
+              {this.state.isGeneratingToken ? (
+                <>
+                  <StripeLogo />
+                  <Spinner
+                    loadingMessage="Validating your card detail..."
```
Comment (paulbrown1982) [inline app/client/components/payment/update/stripeCardInputForm.tsx]: Either remove or pluralise 'detail'.

File: app/client/components/payment/update/currentPaymentDetails.tsx [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
 import { CardDisplay } from "../cardDisplay";
+import { DirectDebitDisplay } from "../directDebitDisplay";
 import { PayPalDisplay } from "../paypalDisplay";
 
 export const CurrentPaymentDetails = (subscription: Subscription) => {
   if (subscription.card) {
     return <CardDisplay {...subscription.card} />;
   } else if (subscription.payPalEmail) {
     return <PayPalDisplay payPalEmail={subscription.payPalEmail} />;
+  } else if (subscription.account) {
```
Comment (paulbrown1982) [inline app/client/components/payment/update/currentPaymentDetails.tsx]: To remove ambiguity, I think this should be bankAccountNumber, else I'd get it confused with Zuora Billing Account. If you don't have control and Subscription is a direct mapping of members-data-api JSON then I guess we can refactor it in future once Digital Pack is also fully migrated over.

File: app/client/components/productPage.tsx [hunk 4]
```diff
@@ -208,41 +201,44 @@ const renderMembershipData = (apiResponse: MembersDataApiResponse) => {
...
-                    "/tier/change"
-                  }
-                >
-                  <Button text="Change tier" />
-                </a>
-              </div>
-            }
-          />
-          <MembershipRow
+          {productType.tierRowLabel ? (
```
Comment (paulbrown1982) [inline app/client/components/productPage.tsx]: I've just realised, that we can hide this button if the tier is Patron!

File: app/shared/productTypes.tsx [hunk 5]
```diff
@@ -1,25 +1,43 @@
 import React from "react";
-import { NoMembership } from "../client/components/cancel/membership/noMembership";
-import { MeValidator } from "../client/components/cancellationFlowWrapper";
+import { MeValidator } from "../client/components/cancel/cancellationFlowWrapper";
+import { NoMembership } from "../client/components/noMembership";
 import { MeResponse } from "./meResponse";
 
-export type ProductName = "membership" | "recurring contribution";
+export type ProductFriendlyName = "membership" | "recurring contribution";
```
Comment (paulbrown1982) [inline app/shared/productTypes.tsx]: Customers might not be thinking of their monthly or annual contribution as a 'recurring contribution'. Maybe raise another ticket to make it contextual to the billing period, then you can use 'monthly' or 'annual'.

File: app/shared/productTypes.tsx [hunk 6]
```diff
@@ -1,25 +1,43 @@
...
-  productName: ProductName;
+  friendlyName: ProductFriendlyName;
+  urlPart: ProductUrlPart;
   validator: MeValidator;
   invalidComponentRenderer: JSX.Element;
+  fetchProductDetail: () => Promise<Response>;
+  tierRowLabel?: string; // no label means row is not displayed
 }
 
+const createProductFetcher = (productApiKeyword: string) => async () =>
```
Comment (paulbrown1982) [inline app/shared/productTypes.tsx]: Could you make the type of this argument a ProductUrlPart?

File: app/client/components/productPage.tsx [hunk 7]
```diff
@@ -264,16 +250,32 @@ const headerCss = css({
...
+    <PageHeaderContainer selectedNavItem={navLinks.contributions}>
+      <h1 className={headerCss}>Contributions</h1>
+    </PageHeaderContainer>
+    <MembersDatApiAsyncLoader
+      fetch={ProductTypes.contributions.fetchProductDetail}
+      render={getProductRenderer(ProductTypes.contributions)}
+      loadingMessage="Loading details of your contributions..."
+    />
+    <PageContainer>
+      <MembershipLinks /> {/*TODO need to have contributions FAQ*/}
```
Comment (jacobwinch) [inline app/client/components/productPage.tsx]: I think MembershipLinks needs to be made generic too?

File: app/shared/meProductResponse.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
+import React from "react";
```
Comment (jacobwinch) [inline app/shared/meProductResponse.ts]: Minor: I would just call this file _productResponse_

File: app/client/components/productPage.tsx [hunk 9]
```diff
@@ -264,16 +250,32 @@ const headerCss = css({
...
+
+export const Contributions = (props: RouteableProps) => (
+  <>
+    <PageHeaderContainer selectedNavItem={navLinks.contributions}>
+      <h1 className={headerCss}>Contributions</h1>
+    </PageHeaderContainer>
+    <MembersDatApiAsyncLoader
+      fetch={ProductTypes.contributions.fetchProductDetail}
+      render={getProductRenderer(ProductTypes.contributions)}
+      loadingMessage="Loading details of your contributions..."
```
Comment (jacobwinch) [inline app/client/components/productPage.tsx]: I'm not sure about this copy... what about: 'Loading your recurring contribution details...'?

(Feel free to ignore if you've already been through this with UX!)

File: app/client/components/payment/update/paymentUpdated.tsx (hunk 1 above)
Comment (twrichards) [reply app/client/components/payment/update/paymentUpdated.tsx]: noted thanks

File: app/shared/productTypes.tsx (hunk 5 above)
Comment (twrichards) [reply app/shared/productTypes.tsx]: will add a TODO in the codebase to cover this

File: app/shared/productTypes.tsx (hunk 6 above)
Comment (twrichards) [reply app/shared/productTypes.tsx]: i've actually reworked this in subsequent PR (https://github.com/guardian/manage-frontend/pull/129)

File: app/client/components/productPage.tsx (hunk 9 above)
Comment (twrichards) [reply app/client/components/productPage.tsx]: i've actually reworked this more generally in subsequent PR (#129)

File: app/client/components/productPage.tsx [hunk 10]
```diff
@@ -218,41 +201,45 @@ const renderMembershipData = (apiResponse: MembersDataApiResponse) => {
...
-                  }
-                >
-                  <Button text="Change tier" />
-                </a>
-              </div>
-            }
-          />
-          <MembershipRow
+          {productType.tierRowLabel &&
+          productType.tierRowLabel.toLowerCase() !== "Patron" ? (
```
Comment (jacobwinch) [inline app/client/components/productPage.tsx]: should this be 'patron'?

File: app/client/components/payment/update/currentPaymentDetails.tsx (hunk 3 above)
Comment (twrichards) [reply app/client/components/payment/update/currentPaymentDetails.tsx]: yes this is to match members data api response, will be improved when we add more DD detail to members data api response

File: app/client/components/productPage.tsx (hunk 4 above)
Comment (twrichards) [reply app/client/components/productPage.tsx]: i did this, but just realised having the change tier button hidden for patron will block the journey to cancellation (via membership site) (as anti-UX as it is) so I will keep the 'Change Tier' for Patron and add a TODO to revisit once we provide a more direct (but still subtle) path to cancellation

**Pull request merged**